%# templat_name['prefix_datasetid','basic_table','fastqc_check','fastqc_table','fastqc_graph','mapping_check','basic_map_table','mappable_ratio_graph','redundant_ratio_graph','peak_summary_table','DHS_ratio_graph','velcro_ratio_graph','venn_graph','correlation_graph','height_distibution_graph','gene_distibution_graph','conservation_graph']
%#  ------Prefix-------prefix_datasetid

%- if RawQC_check
\documentclass[11pt,a4paper]{article}
\usepackage{tabularx}
\usepackage[english]{babel}
\usepackage{array}
\usepackage{graphicx}
\usepackage{color}
\DeclareGraphicsExtensions{.eps,.png,.pdf,.ps}
\begin{document}
\title{dataset \VAR{prefix_datasetid} Chip-Seq QC Report Version 1}
\author {DC group}
\vspace{-1cm}
\maketitle
\abstract{Modern High Throughput experiment analysis always output large amount of information, which sometimes may confuse us. An objective system of assessment throughout all Chip-seq pipeline has been established to
avoid such confusion, Here is a comprehensive report on it.If any question, please contact us.}
\tableofcontents
\setcounter{tocdepth}{2}
\newpage

\section{Reads Genomic Mapping QC measurement}
\begin{quotation}
Modern high throughput sequencers can generate tens of millions of sequences in a single run. Before analysing this sequence to draw biological conclusions you should always perform some simple quality control checks to ensure that the raw data looks good and there are no problems or biases in your data which may affect how you can usefully use it.
\end{quotation}          


%# -------------fastqc--------------fastqc_check : fastqc_table ; fastqc_graph

%- if fastqc_check
\subsection{FastQC Summary table}
\begin{quotation}
FastQC aims to provide a QC report which can spot problems which originate in the sequencer. 
The sequence quality score\footnote{* use 25 as cutoff}  listed in the table allows you to see if a subset of your sequences have universally low quality values. Here we can report the sequence quality scores of the new datasets and use our criteria to judge the quality of the new raw data(Table ~\ref{fastqctable}).\\
\end{quotation}
\begin{table}[h]
\caption{FastQC measurement}\label{fastqctable}
\begin{tabularx}{\textwidth}{ |X|X|X|X| } 
\hline
Sample name & Sequence quality score & QC \\
\hline
\BLOCK{ for line in fastqc_table }
\VAR{line|join(' & ')}
\hline
\BLOCK{ endfor }
\end{tabularx}
\end{table}
\newpage	

\subsection{FastQC score distribution}
We draw the cumulative percentage plot of the sequence quality scores of all historic data and show the sequence quality score of each your new data(Figure ~\ref{fastqcplot}). \newline
\begin{figure}[h]
\setlength{\abovecaptionskip}{0pt} 
\setlength{\belowcaptionskip}{10pt}
\caption{FastQC score distribution plot} \label{fastqcplot}
\centering
{\includegraphics[width=0.9\textwidth]{\VAR{fastqc_graph}}}
\end{figure}
\newpage	
%- endif
%- endif

%# --------mapping QC--------mapping_check : basic_map_table ; mappable_ratio_graph ; redundant_ratio_graph
%# --------basic mapping QC statistics--------

%- if MappingQC_check
\subsection{Basic mapping QC statistics}
\begin{quotation}
Bowtie is an ultrafast, memory-efficient alignment program for aligning short DNA sequence reads to large genomes.\footnote{Langmead B, Trapnell C, Pop M, Salzberg SL: Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome Biology 2009, 10(3):R25}
Short DNA sequence reads sample: AAAGGGCTGAGCTGAATGACTCAT.
Total reads: all reads sequenced in one ChIP-seq experiment.  The file size should be at least 5 mega bytes, or we will fail this QC. 
Mappable reads: reads can align to large genomes when 2 mismatches at most allowed.
Mapping rate: mappable reads / total reads. Percentage of mappable reads. The bigger the percentage is, the better.  
Unique mappable reads: reads that can only map to one location.
Unique mappable locations: locations that can only be mapped by one read.
Unique reads ratio: percentage of unique mappable reads in mappable reads.(Table ~\ref{basicqc})
\end{quotation}
\begin{table}[h]
\caption{Basic QC statistics} \label{basicqc}
\begin{tabularx}{\textwidth}{ |X|X|X|X|X|X|X| } 
\hline
Sample & Total reads & Mappable reads & Mapping rate & Unique Mappable Reads & Unique Mappable location & Redundant rate \\
\hline
\BLOCK{ for line in basic_map_table }
\VAR{line|join(' & ')}
\hline
\BLOCK{ endfor }
\end{tabularx}
\end{table}
\newpage	

%# -------mappable reads ratio---------
\subsection{Mappable reads ratio}
\begin{quotation}
For figure 5, this is a cumulative distribution file of the mappable rates of all the ChIP-seq data in our data collection program,the x axis is the mappable rates and the y axis represents the percentage of this mappable rate among all the data. Given several new sets of data, we can generate the locations of your data among all the data.The mappable rate equals to mapped reads devided by total reads, and the higher this rate is, the better your ChIP-seq experiment you've performed and the higher data quality your data is.(Figure: ~\ref{fig:mappinratio})
\end{quotation}
\begin{figure}[h] 
\caption{Mappable reads ratio} \label{fig:mappinratio}
\setlength{\abovecaptionskip}{0pt} 
\setlength{\belowcaptionskip}{10pt}
\centering
{\includegraphics[width=0.9\textwidth]{\VAR{mappable_ratio_graph}}}
\end{figure}
\newpage

%# -------mappable redundant rate-----------
\subsection{Mappable Redundant rate}
\begin{quotation}
This is a cumulative distribution file of the unique reads ratio of all the ChIP-seq data in our data collection program,the x axis is the unique reads ray and the y axis represents the percentage of this ratio among all the data.Given several new sets of data, we can generate the locations of your data among all the data.The unique here means one read only map to one location of the genome, and the higher this ration is, the higher data quality your data is.(Figure: ~\ref{fig:uni})
\end{quotation}
\begin{figure}[h] 
\caption{Mappable Redundant rate} \label{fig:uni}
\setlength{\abovecaptionskip}{0pt} 
\setlength{\belowcaptionskip}{10pt}
\centering
{\includegraphics[width=0.9\textwidth]{\VAR{redundant_ratio_graph}}}
\end{figure}
\newpage
%- endif


%# ---------peak calling QC---------- peak_summary_table ; DHS_ratio_graph ; velcro_ratio_graph

%- if PeakcallingQC_check
\section{Peak calling QC measurement}
\subsection{Peak calling summary}
\begin{quotation}
This report aims to show the quality of the peaks called by MACS. 
The column of 'Cut off' shows what the Q-value (default is 0.01) is used when processing MACS. And the lower Q-value is used, MACS will call the more confident peaks. 
The 'Peak count' shows the total number of peaks calculated under the condition of a Q-value. 
For the column of 'peaks $\geq$ 20FC', it shows that the number of peaks that require the fold enrichment of tags locate on the genome $\geq$ 20. The log value can give us more confidence in our data.
The 'Shift size' shows that the distance between the Watson or Crick tags enrichment location and the binding site's location.(Table ~\ref{peaksum})
\end{quotation}

\begin{table}[h]
	\caption{Peak summary table} \label{peaksum}
\begin{tabularx}{\textwidth}{ |X|X|X|X|X| } 
\hline
Run name & Cutoff & Peak count & peaks $\geq$ 20FC & Shift size \\
\hline
\BLOCK{ for line in peak_summary_table }
\VAR{line|join(' & ')}
\hline
\BLOCK{ endfor }
\end{tabularx}
\end{table}
\newpage		

%#---------- Hight confident Peak
\subsection{High confident Peak}
\begin{quotation}
The figure shows the distribution for the number's logarithm of peaks that require fold enrichment $\geq$ 10. For zero 10-fold-enrichment peaks, the number is set to 0.1. Here the cutoff is 3. If a data locates on the left side, it won't be considered to have sufficient high-confidence peaks. On the contrary, if a data locates on the right side, it can be thought to have sufficient high-confidence peaks. In other words, if the number of 10-fold enrichment peaks $\geq$ 1000, we can safely say the data have abundant high-confidence peaks.(Figure ~\ref{highconfipeak})
\end{quotation}
\begin{figure}[h] 
\caption{High confident Peak distribution} \label{highconfipeak}
\setlength{\abovecaptionskip}{0pt} 
\setlength{\belowcaptionskip}{10pt}
\centering
{\includegraphics[width=0.8\textwidth]{\VAR{Hight_confident_peak_graph}}}
\end{figure}
\newpage

%#-------Peaks overlapped with DHS----------
\subsection{Peaks overlapped with DHS(Dnase Hypersensitivity sites)}
\begin{quotation}
In eukaryotes, transcription is regulated in a cell-type and condition-specific manner through the association of transcription factors with chromatin. The genome-wide binding sites of transcription factors, or the transcription factor cistromes, are influenced by the active protein levels of the transcription factors, chromatin structure, and DNA sequence.
DNase I hypersensitivity is an alternative measure of chromatin accessibility (Wu 1980). DNase I hypersensitive sites (DHS), short regions of chromatin that are highly sensitive to cleavage by DNase I, typically occur in nucleosome free regions and frequently arise as a result of transcription factor binding. DNase I digestion followed by high-throughput sequencing (DNase-seq) has evolved into a powerful technique for identifying genome-wide DNase hypersensitive sites. \footnote{Ling et al. 2010; John et al. 2011; Siersbaek et al. 2011}
We draw a CDF curve of the overlap with union DHS sites of all DC data, then we can assess the confidence level of  the TF binding sites described by your peak file. (Figure: ~\ref{DHS})
\end{quotation}
\begin{figure}[h]
\begin{center}
\caption{Peaks overlapped with DHS data} \label{DHS}
{\includegraphics[width=0.72\textwidth]{\VAR{DHS_ratio_graph}}}
\end{center}
\end{figure}
\newpage		

%# -------peaks overlaped with Velcro---------

\subsection{Velcro ratio(human only)}
\begin{quotation}
There has a comprehensive set of regions in the human genome that have anomalous, unstructured, high signal/read counts in next gen sequencing experiments independent of cell line and type of experiment. We call them Consensus Signal Artifact Regions (verlcro region). The breadth of cell-lines covered by the ENCODE datasets allows us to accomplish this in a systematic manner.
We use 80 open chromatin tracks (DNase and FAIRE datasets) and 12 ChIP-seq input/control tracks spanning ~60 cell lines in total , identify ~400 Consensus Signal Artifact Regions. We judge the quality and the confidence level of the TF binding sites by its overlap with these region. If the binding sites of a TF have more than 5% binding sites that overlap with these region, it may tell that the TF binding sites are not so reliable.
We draw a CDF curve of the overlap with verlcro region of all DC data, then we can assess the confidence level of  the TF binding sites described by your peak file. (Figure: ~\ref{fig:velcro})
\end{quotation}
\begin{figure}[h] 
\caption{Velcro ratio} \label{fig:velcro}
\setlength{\abovecaptionskip}{0pt} 
\setlength{\belowcaptionskip}{10pt}
\centering
{\includegraphics[width=0.9\textwidth]{\VAR{velcro_ratio_graph}}}
\end{figure}
\newpage

%# -------replicate QC-------replicate_check : venn_graph ; correlation_graph

%- if replicate_check
\subsection{Profile correlation within union peak regions}
\begin{quotation}
If biologist have replicate chipseq experiments,pipline will draw the correlation plot.the score means the correlation among these replicates' profiles. We can judge the replicate experiment quality if the correlaton is high.(Table: ~\ref{fig:profileunion})
\end{quotation}
\begin{figure}[h]
        \caption{Peaks Overlap correlation diagram between Replicates} \label{fig:profileunion}
        \setlength{\abovecaptionskip}{0pt}
        \setlength{\belowcaptionskip}{10pt}
        \centering
        {\includegraphics[width=0.9\textwidth]{\VAR{venn_graph}}}
\end{figure}
\newpage

\subsection{Peaks overlap between Replicates}
\begin{quotation}
If biologist have replicate chipseq experiments,pipline will draw the venn plot.If chipseq peaks  among replicates have more than 1 bp overlap,we take as one overlap.The bigger overlap among replicates, the higher quality in replicate experiments. (Figure: ~\ref{fig:venn})
\end{quotation}
\begin{figure}[h]
        \caption{Peaks Overlap venn diagram between Replicates} \label{fig:venn}
        \setlength{\abovecaptionskip}{0pt}
        \setlength{\belowcaptionskip}{10pt}
        \centering
        {\includegraphics[width=0.9\textwidth]{\VAR{correlation_graph}}}
\end{figure}
\newpage
%- endif
%- endif

%# -------Functional Genomic QC measurement------------height_distibution_graph ; gene_distribution_graph
%#---------Ceas QC---------

%- if AnnotationQC_check
\section{Functional Genomic QC measurement}
\subsection{Peak Height distribution}
\begin{quotation}
Using modern sequencer we do ChIp-seq experiments, thus get millions of factors-binding short reads we grabbed using specific proteins. We get the location and reads density information after  annotate them back to the whole genome, since the reads are located in distinct loci of genome, we can draw a Pie chart to show the distribution of all the reads(i.e.some reads are located in the promoter site while some are in other sites) and its density(namely "Peak Heights").(Figure ~\ref{Peakdist})
\end{quotation}
\begin{figure}[h] 
        \caption{Peak height distribution} \label{Peakdist}
        \setlength{\abovecaptionskip}{0pt} 
        \setlength{\belowcaptionskip}{10pt}
        \centering
        {\includegraphics[width=0.8\textwidth]{\VAR{height_distibution_graph}}}
\end{figure}
\newpage

\subsection{Meta Gene distribution}
\begin{quotation}
Add all the reads density corresponding to its located meta genes(i.e. to the contrast of the whole gene, meta gene indicates the single gene units) together result in those "average profile". Usually there will be a peak near TSS(i.e.transcription start site). Checking those maps gives us an insight about the average reads location near TSS, TTS and the whole genome.(Figure ~\ref{Meta})
\end{quotation}
\begin{figure}[h] 
        \caption{Meta Gene distribution} \label{Meta}
        \setlength{\abovecaptionskip}{0pt} 
        \setlength{\belowcaptionskip}{10pt}
        \centering
        {\includegraphics[width=0.8\textwidth]{\VAR{gene_distribution_graph}}}
\end{figure}
\newpage

%# ---------Conservation QC----------- conservation_graph
\subsection{Peak conservation score} 
still under development.(Figure ~\ref{fig:conservation}) \newline
\begin{figure}[h] 
        \caption{Phascon conservation distribution} \label{fig:conservation}
        \setlength{\abovecaptionskip}{0pt} 
        \setlength{\belowcaptionskip}{10pt}
        \centering
        {\includegraphics[width=0.9\textwidth]{\VAR{conservation_graph}}}
\end{figure}
\newpage
\end{document}

%# -----------motif QC---------------
\subsection{Motif QC measurement analysis}
\begin{quotation}
Sequence motifs are short, recurring patterns in DNA that are presumed to have a biological function. Often they indicate sequence-specific binding sites for proteins such as nucleases and transcription factors (TF). \\footnote{D'haeseleer P: What are DNA sequence motifs? nature biotechnology 2006, 24:423-425.}
We scan the top 1000 peak regions found in MACS to detect the mighty factors' motifs. 
Name: corresponding factor name for motif seqLogo.
Hits: times the motif occurs in the top 1000 regions. More it occurs, more credibility.
Z-score: the value should better be less than -15. And the smaller it is, the more credibility.
Logo: the height represents the reliability and information. If it is very low, then we cannot make sure which base it is.
We might find several motifs different from the antibody we use. For example, we do the AR ChIP-seq experiment, but in the motif finding results we might also find FoxA1, suggest that FoxA1 might be a co-factor for AR. 
In order to avoid redundant motif prediction and motifs with too much similarity between denovo and past identified ones, we used a  function to merge the replicates and redundant. 
We search in cistrome database and use denovo motif finding methods 
to generate prediction of experiment factor's motifs. 
In order to avoid redundant motif prediction and motifs with too much similarity between denovo and past identified ones, we used a  function to merge the replicates and redundant. (Table ~\\ref{motif})
\end{quotation}
\newpage

\begin{table}[h]
        \caption{Seqpos QC measurement} \label{motif}
        \newcolumntype{R}{>{\\raggedleft\\arraybackslash}X}%%
\begin{tabularx}{1.2\\textwidth}{ |X|R|X|X|X| }
\hline
factor name & Hits & Z-score & motif logo \\tabularnewline
\hline
\BLOCK{ for line in motif_table }
\VAR{line|join(' & ')}
\hline
\BLOCK{ endfor }
\end{tabularx}
\end{table}
\end{document}
%- endif